test/test_centroids.py

```diff
@@ -0,0 +1,49 @@
+import os
+from unittest import TestCase
+import numpy as np
+import numpy.testing as nt
+import uxarray as ux
+from pathlib import Path
+from uxarray.grid.coordinates import _populate_centroid_coord
+
+current_path = Path(os.path.dirname(os.path.realpath(__file__)))
+
+gridfile_CSne8 = current_path / "meshfiles" / "scrip" / "outCSne8" / "outCSne8.nc"
+
+
+class TestCentroids(TestCase):
+
+    def test_centroids_from_mean_verts(self):
```

Comment:
Please include testcases that are more complicate:

1. A polygon that has more than 3 vertexes
2. A polygon that one of its vertexes are far away from the other
3. Other irregular shape polygon

Reply:
I would say that our test `test_centroids_from_mean_verts_scrip` is pretty comprehensive, since it runs the algorithm on a dataset that already has `Mesh2_face_x/y` and the results match up. This dataset is also made up of quads, which checks off the first point.

Reply:
> I would say that our test `test_centroids_from_mean_verts_scrip` is pretty comprehensive, since it runs the algorithm on a dataset that already has `Mesh2_face_x/y` and the results match up. This dataset is also made up of quads, which checks off the first point.
> 
> However, additional tests would be good to have.

So does this SCRIP dataset use a real spherical surface presentation (cruved) or an approximate one using triangles?? 

The point is that only averaging the coordinates without normalization (our current implementation), the centroid is not actually on the sphere (Unless that's exactly what we want)

Reply:
> > I would say that our test `test_centroids_from_mean_verts_scrip` is pretty comprehensive, since it runs the algorithm on a dataset that already has `Mesh2_face_x/y` and the results match up. This dataset is also made up of quads, which checks off the first point.
> > However, additional tests would be good to have.
> 
> So does this SCRIP dataset use a real spherical surface presentation (curved) or an approximate one using triangles?
> 
> The point is that only averaging the coordinates without normalization (our current implementation), the centroid is not actually on the sphere (Unless that's exactly what we want)

I believe so

*[SCRIP](https://github.com/SCRIP-Project/SCRIP): A Spherical Coordinate Remapping and Interpolation Package (SCRIP) is a software package used to generate interpolation weights for remapping fields from one grid to another in spherical geometry.  It was developed at the Los Alamos National Laboratory by Philip W. Jones. The SCRIP library was converted from Fortran to ANSI C and is used as the base for remapping operators in CDO.

Reply:
The SCRIP test doesn’t test the accuracy well, because if latlon already exists it doesn’t recreate it using our definition. It just assigns it to what it already is. Really all it does is convert it to get the Cartesian. This is what I was saying about repopulation not doing anything in that regard.

Reply:
> The SCRIP test doesn’t test the accuracy well, because if latlon already exists it doesn’t recreate it using our definition. It just assigns it to what it already is. Really all it does is convert it to get the Cartesian. This is what I was saying about repopulation not doing anything in that regard.

I might be missing something, but are you sure this is the case? I've stepped through the execution of the test case and the population function it does indeed compute `centroid_lon` and `centroid_lat` and then re-populates the data array.

#### `Mesh2_face_x/y` and  `_populate_centroid_coord(uxgrid, repopulate=True)` is called

##### Computation
```Python
# block (1)
    if "Mesh2_face_x" not in self._ds or repopulate:
        # Construct the centroids if there are none stored
        if "Mesh2_face_cart_x" not in self._ds:
            centroid_x, centroid_y, centroid_z = _construct_xyz_centroids(
                node_x, node_y, node_z, face_nodes, nNodes_per_face)

        else:
            # If there are cartesian centroids already use those instead
            centroid_x, centroid_y, centroid_z = self.Mesh2_face_cart_x, self.Mesh2_face_cart_y, self.Mesh2_face_cart_z
# block (2)
        # Convert from xyz to latlon
        centroid_lon, centroid_lat = _get_lonlat_from_xyz(
            centroid_x, centroid_y, centroid_z)
# block (3)
    else:
        # Convert to xyz if there are latlon centroids already stored
        centroid_lon, centroid_lat = self.Mesh2_face_x.values, self.Mesh2_face_y.values
        centroid_x, centroid_y, centroid_z = _get_xyz_from_lonlat(
            centroid_lon, centroid_lat)
```

We enter block (1) since `repopulate` is set to True, which computes our `centroid_x/y/z`

Block (2) is followed and we computed out `centroid_lat/lon` (recomputed, not the original `Mesh2_face_x/y`

We never actually enter block (3)


Any thoughts?

Reply:
Oh yeah, you are right, somehow I forgot about the or statement there, my bad! That’s actually really good then.

Reply:
> Oh yeah, you are right, somehow I forgot about the or statement there, my bad! That’s actually really good then.

All good!

Reply:
> > > I would say that our test `test_centroids_from_mean_verts_scrip` is pretty comprehensive, since it runs the algorithm on a dataset that already has `Mesh2_face_x/y` and the results match up. This dataset is also made up of quads, which checks off the first point.
> > > However, additional tests would be good to have.
> > 
> > 
> > So does this SCRIP dataset use a real spherical surface presentation (curved) or an approximate one using triangles?
> > The point is that only averaging the coordinates without normalization (our current implementation), the centroid is not actually on the sphere (Unless that's exactly what we want)
> 
> I believe so
> 
> *[SCRIP](https://github.com/SCRIP-Project/SCRIP): A Spherical Coordinate Remapping and Interpolation Package (SCRIP) is a software package used to generate interpolation weights for remapping fields from one grid to another in spherical geometry. It was developed at the Los Alamos National Laboratory by Philip W. Jones. The SCRIP library was converted from Fortran to ANSI C and is used as the base for remapping operators in CDO.

A spherical coordinate system doesn't necessarily mean they are using a **curve** surface here. Many application utilizes triangles to approximate the sphere, which works fine as well, it's just not our case.

Again, in the geometry calculation, directly averaging the cartesian coordinates of a **curved** "triangle" in the 3-D spaces, the resulted "centroid" shouldn't be on the spherical surface. 

However, the normalization process can project this "centroid" to the surface. But I am not sure if this "projected" centroid is the real centroid according to the definition.

The curved surface is way more complicated than the normal plane surface. And it's understandable that we want to stay with the current implementation to expedite the process as well. Just make sure to clarify that in the comment.